### PR TITLE
Defining SQLite primary keys inline with the column as "INTEGER PRIMA…

### PR DESCRIPTION
**Pull request by icnocop**
…RY KEY AUTOINCREMENT"

### CODE REVIEW

**Comment by icnocop** [comment]
This change is specifically for SQLite. For example, when converting an MsSQL database to SQLite.

The change makes primary keys have the `AUTOINCREMENT` keyword.

For example:
Before these code changes:
```
CREATE TABLE [Table1] (
  [Column1] INTEGER NOT NULL
, [Column2] bit DEFAULT 0 NOT NULL
, CONSTRAINT [PK_Table1] PRIMARY KEY ([Column1])
);
```

After these code changes:
```
CREATE TABLE [Table1] (
  [Column1] INTEGER PRIMARY KEY AUTOINCREMENT NOT NULL
, [Column2] bit DEFAULT 0 NOT NULL
);
```

https://www.sqlite.org/autoinc.html

> If the AUTOINCREMENT keyword appears after INTEGER PRIMARY KEY, that changes the automatic ROWID assignment algorithm to prevent the reuse of ROWIDs over the lifetime of the database. In other words, the purpose of AUTOINCREMENT is to prevent the reuse of ROWIDs from previously deleted rows. 

This behavior is more inline with the definition of the table and behavior in MsSQL:
```
CREATE TABLE [Table1] (
    [Column1]                    BIGINT         IDENTITY (1, 1) NOT NULL,
    [Column2]                    BIT            CONSTRAINT [DF_Table1_Column2] DEFAULT ((0)) NOT NULL,
    CONSTRAINT [PK_Table1] PRIMARY KEY CLUSTERED ([Column1] ASC),
);
```
https://docs.microsoft.com/en-us/sql/t-sql/statements/create-table-transact-sql-identity-property

> Reuse of values - For a given identity property with specific seed/increment, the identity values are not reused by the engine. If a particular insert statement fails or if the insert statement is rolled back then the consumed identity values are lost and will not be generated again. This can result in gaps when the subsequent identity values are generated.

So basically, I don't expect the generated SQLite database to re-use primary key values from rows that were deleted.
This becomes an issue if the database keeps an audit trail of changes and references primary keys for example. If primary keys are re-used, then the audit trail becomes inaccurate.

**Comment by ErikEJ** [comment]
But I read this (from your link)

The AUTOINCREMENT keyword imposes extra CPU, memory, disk space, and disk I/O overhead and **should be avoided if not strictly needed.** It is usually not needed. 

In SQLite, a column with type **INTEGER PRIMARY KEY** is an alias for the ROWID (except in WITHOUT ROWID tables) which is always a 64-bit signed integer.

**Comment by icnocop** [comment]
Should the `AUTOINCREMENT` keyword be added optionally?
Should it be enabled by default?

**Comment by ErikEJ** [comment]
What happens currently when you export a SQL Server table with an IDENTITY column?

**Comment by icnocop** [comment]
I just pushed a commit with the following test:
```
        [Test]
        public void GenerateDatabaseScript_FromMsSqlToSqlite_IdentityColumn()
        {
            string createTableQuery =
                @"CREATE TABLE [Test]
                (
                        [TestId]  BIGINT         IDENTITY NOT NULL,
                        [Column1] nvarchar(256)  NULL
                );";

            string expectedSql = string.Join(Environment.NewLine, new[]
            {
                "CREATE TABLE [Test] (",
                "  [TestId] INTEGER NOT NULL",
                ", [Column1] nvarchar(256) NULL COLLATE NOCASE",
                ")"
            });

            AssertMsSqlToSqliteGeneratorContains(createTableQuery, "Test", expectedSql);
        }
```

**Comment by ErikEJ** [comment]
What if you declare the I'd as primary key?

**Comment by ErikEJ** [comment]
Hmmmm… The second and third tests seem to be OK. The question is what should happen, if the MSSQL column is defined a an identity column, but not being a primary key?

 Maybe we can ask @ericsink fro advice? (Zumero)

**Comment by ericsink** [comment]
SQLite can’t do an identity/autoincrement column that is not a primary key. But you can implement something similar using an extra table and triggers.

**Comment by ErikEJ** [comment]
I wonder if anything actually needs to change - the issue is that you can create a SQL table like this:

```
CREATE TABLE [Test]
                (
                        [TestId]  BIGINT         IDENTITY NOT NULL,
                        [Column1] nvarchar(256)  NULL,
                       [Key] int PRIMARY KEY NOT NULL
                );
```

**Comment by icnocop** [comment]
I'm not exactly sure what you mean.

That doesn't create an SQLite table with `[Key] INTEGER PRIMARY KEY AUTOINCREMENT NOT NULL`.

Given that MsSQL table, what do expect the SQLite table to be generated as?

The changes in this PR may not support all cases, but it supports the simplest (any maybe the most common) case as in https://github.com/ErikEJ/SqlCeToolbox/pull/763#issuecomment-447553131.

**Comment by ErikEJ** [comment]
Sorry, I misread your tests - looks good!
